contracts/MarketOracle.sol

```diff
@@ -3,149 +3,208 @@ pragma solidity 0.4.24;
-     *      24hrs, then there is effectively no exchange rate and that value should be ignored by
-     *      the client.
-     * @return exchangeRate: Volume weighted average of exchange rates.
-     *         volume: Total trade volume of the last reported 24 hours in Token volume.
-     */
-    function getPriceAnd24HourVolume()
-        external
-        returns (uint256, uint256)
-    {
-        uint256 volumeWeightedSum = 0;
-        uint256 volumeSum = 0;
-        uint256 partialRate = 0;
-        uint256 partialVolume = 0;
-        bool isSourceFresh = false;
-
-        for (uint256 i = 0; i < whitelist.length; i++) {
-            (isSourceFresh, partialRate, partialVolume) = whitelist[i].getReport();
-
-            if (!isSourceFresh) {
-                emit LogSourceExpired(whitelist[i]);
-                continue;
-            }
+    // Reports indexed by provider address. Report[0].timestamp > 0
+    // indicates provider existence.
+    mapping (address => Report[2]) public providerReports;
 
-            volumeWeightedSum = volumeWeightedSum.add(partialRate.mul(partialVolume));
-            volumeSum = volumeSum.add(partialVolume);
-        }
+    event ProviderAdded(address provider);
+    event ProviderRemoved(address provider);
+    event ReportTimestampOutOfRange(address provider);
+
+    // The number of seconds after which the report is deemed expired.
+    uint256 public reportExpirationTimeSec;
+
+    // The number of seconds since reporting that has to pass before a report
+    // is usable.
+    uint256 public reportDelaySec;
 
-        // No explicit fixed point normalization is done as dividing by volumeSum normalizes
-        // to exchangeRate's format.
-        uint256 exchangeRate = volumeSum > 0
-            ? volumeWeightedSum.div(volumeSum)
-            : 0;
-        return (exchangeRate, volumeSum);
+    uint256 public minimumProviders = 1;
+
+    // Timestamp of 1 is used to mark uninitialized and invalidated data.
+    // This is needed so that timestamp of 1 is always considered expired.
+    uint256 private constant MAX_REPORT_EXPIRATION_TIME = 10 years;
+
+    constructor(uint256 reportExpirationTimeSec_,
+                uint256 reportDelaySec_,
+                uint256 minimumProviders_)
+        public
+    {
+        require(reportExpirationTimeSec_ <= MAX_REPORT_EXPIRATION_TIME);
+        require(minimumProviders_ > 0);
+        reportExpirationTimeSec = reportExpirationTimeSec_;
+        reportDelaySec = reportDelaySec_;
+        minimumProviders = minimumProviders_;
     }
 
-    /**
-     * @dev Adds a market source to the whitelist.
-     * Upgradeable contracts should never be added,
-     * because the logic could be changed after the whitelisting process.
-     * @param source Address of the MarketSource.
-     */
-    function addSource(MarketSource source)
+    function setReportExpirationTimeSec(uint256 reportExpirationTimeSec_)
```

Comment:
Still need some NatSpec comments here. (and others)

Reply:
Done.